# Tugas Minggu ke-7 Pembelajaran Mesin
| Name           | NRP        | Kelas     |
| ---            | ---        | ----------|
| Algof Kristian Zega | 5025231235 | Pembelajaran Mesin (D) |

**Source Kaggle** <br>
[https://www.kaggle.com/competitions/hands-on-2-pembelajaran-mesin-2025](https://www.kaggle.com/competitions/hands-on-2-pembelajaran-mesin-2025/overview)

## 1. Problem Definition

### 1.1 Overview

This dataset simulates user behavior on an e-commerce website using Google Analytics session-level data. Each row represents a unique session, with various behavioral, traffic source, and device-related attributes. The goal is to predict whether a user will make a purchase on a return visit, based on information from their current session.

### 1.2 Objective

Your task is to build a model that predicts the binary outcome:

Will the user make a purchase if they return in a later session?

This can help businesses understand visitor intent, personalize user experience, and retarget high-quality traffic.

## 2. Data Collection

Download dataset dari Kaggle
```sh
kaggle competitions download -c hands-on-2-pembelajaran-mesin-2025
```

In [1]:
import pandas as pd

train_df = pd.read_csv("train.csv")
train_df

,id,unique_session_id,bounces,time_on_site,pageviews,hits,session_quality_dim,latest_ecommerce_progress,avg_time_per_page,source,medium,channelGrouping,deviceCategory,operatingSystem,browser,country,city,will_buy_on_return_visit
0,0,98463921871485068561470319157,0,2818,68,86,0,6,41.992188,mall.googleplex.com,referral,referral,desktop,macintosh,chrome,canada,toronto,1
1,1,64969931561991723631486389718,0,32,4,4,0,0,8.000000,(direct),NaN,direct,desktop,linux,uc browser,india,kharagpur,1
2,2,2117346273301973411473489944,0,93,8,11,0,2,11.625000,mail.google.com,referral,referral,desktop,chrome os,chrome,united states,san francisco,1
3,3,8934726040229318731481512421,0,265,15,21,0,2,17.333333,lunametrics.com,referral,referral,mobile,android,firefox,australia,sydney,0
4,4,5054831196771327751486859045,1,0,1,1,0,0,NaN,youtube.com,referral,social,desktop,windows,internet explorer,thailand,bangkok,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3995,4992,97935044993066181501482069964,1,0,1,1,0,0,NaN,(direct),NaN,direct,mobile,android,opera mini,burkina faso,ouagadougou,0
3996,4993,50367472411835115651495891843,1,0,1,1,0,0,NaN,google,cpc,paid search,tablet,android,chrome,india,hyderabad,0
3997,4994,2079573967163814741473478241,0,50,5,5,0,0,10.000000,mall.googleplex.com,referral,referral,desktop,linux,chrome,canada,kitchener,1
3998,4995,09127428276670369691481539217,0,28,3,4,0,2,9.333333,sites.google.com,referral,referral,desktop,linux,chrome,united states,mountain view,1


In [2]:
test_df = pd.read_csv("test.csv")
test_df

,id,unique_session_id,bounces,time_on_site,pageviews,hits,session_quality_dim,latest_ecommerce_progress,avg_time_per_page,source,medium,channelGrouping,deviceCategory,operatingSystem,browser,country,city
0,44,70254118674433699171499787955,0,10,3,3,2,0,3.333333,mall.googleplex.com,referral,referral,desktop,windows,chrome,united states,new york
1,51,13942618714780242461487108189,0,72,9,11,0,4,8.000000,mall.googleplex.com,referral,referral,desktop,macintosh,chrome,canada,toronto
2,60,09076406666416484141488339781,1,0,1,1,0,0,NaN,youtube.com,referral,social,tablet,android,chrome,vietnam,ho chi minh city
3,70,15714003792776280801481560845,1,0,1,1,0,0,NaN,analytics.google.com,referral,referral,tablet,ios,safari,vietnam,hanoi
4,77,92369863251057752251482101596,0,148,9,11,0,5,16.333333,(direct),NaN,direct,desktop,macintosh,chrome,canada,toronto
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,4982,53271684582812447671480397009,1,0,1,1,0,0,NaN,google,organic,organic search,desktop,windows,chrome,hong kong,NaN
996,4985,45355569263733749281487924598,0,768,17,21,0,6,46.058824,sites.google.com,referral,referral,desktop,macintosh,chrome,united states,mountain view
997,4996,12416167441564751191480326296,0,320,21,24,0,6,14.090909,dfa,cpm,display,desktop,macintosh,chrome,united states,new york
998,4997,96710519571151703481485273816,0,23,2,2,0,0,11.500000,partners,affiliate,affiliates,desktop,windows,firefox,sweden,stockholm


## 3. Exploratory Data Analysis

### 3.1 Feature

- **unique_session_id**  
  Unique identifier for each session (based on fullVisitorId and visitId).
  
- **bounces**  
  Indicates if the session ended with only one pageview.
  
- **time_on_site**  
  Total time spent on the site during the session (in seconds).
  
- **pageviews**  
  Number of pages viewed during the session.
  
- **hits**  
  Total number of user interactions during the session.
  
- **session_quality_dim**  
  Proprietary quality score (1–100) estimating purchase intent.
  
- **latest_ecommerce_progress**  
  The furthest stage reached in the e-commerce funnel.
  
- **avg_time_per_page**  
  Average time per page (calculated as time_on_site / pageviews).
  
- **source**  
  The source of the traffic (e.g., google, bing, direct).
  
- **medium**  
  The medium of the traffic (e.g., organic, cpc, referral).
  
- **channelGrouping**  
  High-level channel category (e.g., Referral, Organic Search, Direct).
  
- **deviceCategory**  
  The category of the device used (desktop, tablet, mobile).
  
- **operatingSystem**  
  The category of the device operating system used.
  
- **browser**  
  The category of the device browser used.
  
- **country**  
  Country from where the user accessed the site.
  
- **city**  
  City from where the user accessed the site.

Mencari missing value dari dataset

In [ ]:
train_df.info()

In [ ]:
train_df.describe()

In [ ]:
train_df.isna().sum()

## 4. Data Preprocessing

Mean imputation untuk feature `avg_time_per_page`

In [14]:
train_df['avg_time_per_page'].fillna(train_df['avg_time_per_page'].mean(), inplace=True)

C:\Users\Algof Kristian Zega\AppData\Local\Temp\ipykernel_17508\1231506095.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  train_df['avg_time_per_page'].fillna(train_df['avg_time_per_page'].mean(), inplace=True)


Most frequent imputation untuk feature `operatingSystem`

In [15]:
train_df['operatingSystem'].mode()[0]

'macintosh'

In [16]:
train_df['operatingSystem'].fillna('machintosh', inplace=True)

C:\Users\Algof Kristian Zega\AppData\Local\Temp\ipykernel_17508\4149413241.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  train_df['operatingSystem'].fillna('machintosh', inplace=True)


Drop kolom feature `city` dan `medium` karena terlalu banyak missing value

In [17]:
train_df.drop(['city', 'medium'], axis=1, inplace=True)

Drop kolom `id` dan `unique_session_id` karena tidak terlalu berpengaruh

In [18]:
train_df.drop(['unique_session_id', 'id'], axis=1, inplace=True)

*Meng-encode* fitur string ke numerik agar bisa dibaca oleh model

In [19]:
from sklearn.preprocessing import LabelEncoder

kolom_encode = ['source', 'channelGrouping', 'deviceCategory', 'operatingSystem', 'browser', 'country']

le_dict = {}

for kolom in kolom_encode:
    le = LabelEncoder()
    train_df[kolom] = le.fit_transform(train_df[kolom])
    le_dict[kolom] = le

In [ ]:
# del kolom, le, le_dict, kolom_encode

---

Data Test Preprocessing

Most frequent imputation untuk feature `operatingSystem`

In [7]:
train_df['operatingSystem'].mode()[0]

'macintosh'

In [8]:
test_df['operatingSystem'] = test_df['operatingSystem'].fillna('macintosh')

Mean imputation untuk feature `avg_time_per_page`

In [4]:
train_df['avg_time_per_page'].mean()

np.float64(34.464465276617474)

In [5]:
test_df['avg_time_per_page'] = test_df['avg_time_per_page'].fillna(34.464465276617474)

Drop kolom feature `city` dan `medium` karena terlalu banyak missing value

In [10]:
test_df.drop(['city', 'medium'], axis=1, inplace=True)

Drop kolom `id` dan `unique_session_id` karena tidak terlalu berpengaruh

In [12]:
test_df.drop(['unique_session_id', 'id'], axis=1, inplace=True)

*Meng-encode* fitur string ke numerik agar bisa dibaca oleh model

In [45]:
from sklearn.preprocessing import LabelEncoder

kolom_encode = ['source', 'channelGrouping', 'deviceCategory', 'operatingSystem', 'browser', 'country']

for kolom in kolom_encode:
    le = LabelEncoder()
    test_df[kolom] = le.fit_transform(test_df[kolom])

## 4. Model Selection

**🔹 Model Klasik (Tradisional)**
- Logistic Regression  
- K-Nearest Neighbors (KNN)  
- Naive Bayes  
  - Gaussian Naive Bayes  
  - Multinomial Naive Bayes  
  - Bernoulli Naive Bayes  
- Decision Tree  
- Random Forest  
- Support Vector Machine (SVM)  
- Linear Discriminant Analysis (LDA)  
- Quadratic Discriminant Analysis (QDA)  
- Perceptron (linear classifier)  

**🔹 Ensemble Methods**

**Bagging Classifier**
- Misalnya: Bagged Decision Trees  
- Random Forest (juga termasuk ensemble)  

**Boosting Algorithms**
- AdaBoost  
- Gradient Boosting  
- XGBoost  
- LightGBM  
- CatBoost  

**Stacking & Voting**
- Stacking Classifier  
- Voting Classifier

### 4.1 K-Nearest Neighbour

In [23]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

numeric_feature = ['bounces', 'time_on_site', 'pageviews', 'hits', 'session_quality_dim',
                   'latest_ecommerce_progress', 'avg_time_per_page']

X = train_df[numeric_feature] # fitur (features)
y = train_df['will_buy_on_return_visit'] # target (labels)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

"""
for k in range (1, 20): 
	knn = KNeighborsClassifier(n_neighbors=k)
	knn.fit(X_train, y_train)
	y_pred = knn.predict(X_test)
	accuracy = accuracy_score(y_test, y_pred)
	print(f"Akurasi model KNN: {accuracy * 100:.2f}%")
"""
knn = KNeighborsClassifier(n_neighbors=8)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Akurasi model KNN: {accuracy * 100:.2f}%")

Akurasi model KNN: 75.00%


### 4.2 Decision Tree

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

X = train_df.drop(['will_buy_on_return_visit'], axis=1)  # Fitur
y = train_df['will_buy_on_return_visit']  # Target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

model = DecisionTreeClassifier(random_state=42)

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

In [ ]:
# del X, y, X_train, X_test, y_train, y_test, model, y_pred

In [ ]:
from sklearn.tree import plot_tree
import matplotlib.pyplot as plt

# Visualisasi Decision Tree
plt.figure(figsize=(12,8))
plot_tree(model, filled=True, feature_names=X.columns, class_names=[str(i) for i in model.classes_], rounded=True)
plt.show()

### 4.3 Cat Boost

In [24]:
from sklearn.model_selection import train_test_split

X = train_df.drop(['will_buy_on_return_visit'], axis=1)  # Fitur
y = train_df['will_buy_on_return_visit']  # Target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [66]:
from catboost import CatBoostClassifier

model = CatBoostClassifier(iterations=1000,  # Jumlah pohon (trees)
                           learning_rate=0.05,  # Learning rate
                           depth=6,  # Kedalaman pohon
                           verbose=100)  # Tampilkan progress setiap 100 iterasi

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

0:	learn: 0.6482921	total: 150ms	remaining: 2m 30s
100:	learn: 0.2172385	total: 323ms	remaining: 2.88s
200:	learn: 0.1770546	total: 508ms	remaining: 2.02s
300:	learn: 0.1500512	total: 693ms	remaining: 1.61s
400:	learn: 0.1302479	total: 878ms	remaining: 1.31s
500:	learn: 0.1110795	total: 1.06s	remaining: 1.05s
600:	learn: 0.0981681	total: 1.25s	remaining: 832ms
700:	learn: 0.0878499	total: 1.45s	remaining: 619ms
800:	learn: 0.0786922	total: 1.64s	remaining: 407ms
900:	learn: 0.0719822	total: 1.82s	remaining: 200ms
999:	learn: 0.0663899	total: 2.01s	remaining: 0us
Accuracy: 91.33%


In [68]:
test_pred = model.predict(test_df)

df_test_pred = pd.DataFrame(test_pred)

df_test_pred.to_csv(f"submission-{3}.csv", index=False, header=['will_buy_on_return_visit'])

### 4.4 Gaussian Naive Bayes

🧠 Untuk fitur numerik yang diasumsikan berdistribusi normal/Gaussian.

In [ ]:
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB
from sklearn.metrics import accuracy_score

model = GaussianNB()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

In [ ]:
del model, y_pred

### 4.5 Multinomial Naive Bayes

🧠 Untuk count data (misal: jumlah kata pada dokumen). Cocok banget buat NLP / text classification.

In [ ]:
model = MultinomialNB()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

In [ ]:
del model, y_pred, accuracy

⚠️ Note:

* Fitur harus non-negatif (biasanya integer ≥ 0).

* Cocok untuk data diskrit.

### 4.6 Bernoulli Naive Bayes

🧠 Untuk fitur biner (0 atau 1), misalnya kehadiran/ketiadaan fitur tertentu.

In [ ]:
model = BernoulliNB()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

In [ ]:
del model, y_pred, accuracy

⚠️ Pastikan data fitur sudah dalam bentuk biner

### 4.7 Random Forest Classifier

In [73]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score

model = RandomForestClassifier(n_estimators=100, random_state=42, max_depth=10, min_samples_split=10)
# n_estimators = jumlah pohon decision tree
# Best Parameters: {'max_depth': 10, 'min_samples_split': 10, 'n_estimators': 100}

# Melakukan cross-validation 5-fold dan menghitung skor akurasi
scores = cross_val_score(model, X, y, cv=5)  # cv=5 artinya 5-fold cross-validation

# Menampilkan hasil skor akurasi untuk tiap fold
print(f"Cross-validation Scores: {scores}")

# Menghitung dan menampilkan average score
average_score = scores.mean()
print(f"Average Cross-Validation Score: {average_score}")
"""
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")
"""

Cross-validation Scores: [0.885   0.8875  0.90375 0.8975  0.8875 ]
Average Cross-Validation Score: 0.89225


'\nmodel.fit(X_train, y_train)\n\ny_pred = model.predict(X_test)\n\naccuracy = accuracy_score(y_test, y_pred)\nprint(f"Accuracy: {accuracy * 100:.2f}%")\n'

In [72]:
test_pred = model.predict(test_df)

df_test_pred = pd.DataFrame(test_pred)

df_test_pred.to_csv(f"submission-{4}.csv", index=False, header=['will_buy_on_return_visit'])

In [ ]:
del model, y_pred, accuracy

RandomForest dengan optimasi GridSearch

In [69]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

model = RandomForestClassifier(n_estimators=100, random_state=42)
# n_estimators = jumlah pohon decision tree

param_grid = {
    'n_estimators': [50, 100, 200],  # Jumlah pohon decision tree
    'max_depth': [10, 20, None],      # Kedalaman pohon
    'min_samples_split': [2, 5, 10]   # Minimum sampel untuk split
}

grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='accuracy')

grid_search.fit(X_train, y_train)

print(f"Best Parameters: {grid_search.best_params_}")

best_model = grid_search.best_estimator_

y_pred = best_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

Best Parameters: {'max_depth': 10, 'min_samples_split': 10, 'n_estimators': 100}
Accuracy: 89.42%


In [ ]:
del model, param_grid, grid_search, best_model, y_pred, accuracy

### 4.8 Ada Boost

In [76]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

model = AdaBoostClassifier(n_estimators=100,  # Jumlah iterasi atau jumlah pohon yang digunakan
                           learning_rate=1.7,  # Learning rate : 0.8 1.6 1.7 1.8
                           random_state=42)

scores = cross_val_score(model, X, y, cv=5)  # cv=5 artinya 5-fold cross-validation

# Menampilkan hasil skor akurasi untuk tiap fold
print(f"Cross-validation Scores: {scores}")

# Menghitung dan menampilkan average score
average_score = scores.mean()
print(f"Average Cross-Validation Score: {average_score}")

# model.fit(X_train, y_train)

# y_pred = model.predict(X_test)

# accuracy = accuracy_score(y_test, y_pred)
# print(f"Accuracy: {accuracy * 100:.2f}%")

Cross-validation Scores: [0.8925  0.9025  0.89    0.905   0.90125]
Average Cross-Validation Score: 0.89825


In [34]:
del model, y_pred, accuracy

### 4.9 XG Boost

In [75]:
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix

model = xgb.XGBClassifier(
    objective='binary:logistic',  # Untuk klasifikasi biner
    eval_metric='logloss',        # Metode evaluasi log loss
    use_label_encoder=False,       # Menonaktifkan label encoder yang deprecated
	#  {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 200}
    learning_rate=0.05,
    max_depth=3,
    n_estimators=300
)

scores = cross_val_score(model, X, y, cv=5)  # cv=5 artinya 5-fold cross-validation

# Menampilkan hasil skor akurasi untuk tiap fold
print(f"Cross-validation Scores: {scores}")

# Menghitung dan menampilkan average score
average_score = scores.mean()
print(f"Average Cross-Validation Score: {average_score}")

# model.fit(X_train, y_train)

# y_pred = model.predict(X_test)

# accuracy = accuracy_score(y_test, y_pred)
# print(f"Akurasi: {accuracy * 100:.2f}%")

# cm = confusion_matrix(y_test, y_pred)
# print("Confusion Matrix:")
# print(cm)

# xgb.plot_importance(model, importance_type='weight')

c:\Users\Algof Kristian Zega\AppData\Local\Programs\Python\Python312\Lib\site-packages\xgboost\training.py:183: UserWarning: [17:02:49] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
c:\Users\Algof Kristian Zega\AppData\Local\Programs\Python\Python312\Lib\site-packages\xgboost\training.py:183: UserWarning: [17:02:49] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
c:\Users\Algof Kristian Zega\AppData\Local\Programs\Python\Python312\Lib\site-packages\xgboost\training.py:183: UserWarning: [17:02:49] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
c:\Users\Algof Kristian Zega\AppData\Local\Programs\Python\Python312\Lib\site-packages\xgboost\tr

Cross-validation Scores: [0.89625 0.9075  0.90875 0.91    0.9    ]
Average Cross-Validation Score: 0.9045


In [ ]:
del model, y_pred, accuracy#, cm

Optimasi XGBoost

In [ ]:
"""
from sklearn.model_selection import GridSearchCV
import numpy as np

model = xgb.XGBClassifier(use_label_encoder=False, objective='binary:logistic')

param_grid = {
    'max_depth': range(3,10),
    'learning_rate': np.arange(0.01, 2.0, 0.1),
    'n_estimators': np.arange(100, 400, 50)
}

grid_search = GridSearchCV(estimator=model, param_grid=param_grid, scoring='accuracy', cv=5, verbose=1)
grid_search.fit(X_train, y_train)

print("Best parameters found: ", grid_search.best_params_)

# {'learning_rate': np.float64(0.21000000000000002), 'max_depth': 3, 'n_estimators': np.int64(100)}
best_model = grid_search.best_estimator_

y_pred = best_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Akurasi model terbaik: {accuracy * 100:.2f}%")
"""

### 4.10 Gradient Boosting

In [74]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score

gb_model = GradientBoostingClassifier(
    n_estimators=200,  # Jumlah estimator (pohon)
    learning_rate=0.05,  # Kecepatan pembelajaran
    max_depth=4,  # Kedalaman maksimum pohon
    subsample=0.6  # Menggunakan 60% sampel data untuk melatih setiap pohon
)

scores = cross_val_score(gb_model, X, y, cv=5)  # cv=5 artinya 5-fold cross-validation

# Menampilkan hasil skor akurasi untuk tiap fold
print(f"Cross-validation Scores: {scores}")

# Menghitung dan menampilkan average score
average_score = scores.mean()
print(f"Average Cross-Validation Score: {average_score}")

# gb_model.fit(X_train, y_train)

# y_pred = gb_model.predict(X_test)

# accuracy = accuracy_score(y_test, y_pred)
# print(f"Akurasi model: {accuracy * 100:.2f}%")

Cross-validation Scores: [0.8975 0.9075 0.9075 0.9175 0.8975]
Average Cross-Validation Score: 0.9055


In [46]:
test_pred = gb_model.predict(test_df)

df_test_pred = pd.DataFrame(test_pred)

df_test_pred.to_csv(f"submission-{1}.csv", index=False, header=False)

In [61]:
del test_pred, df_test_pred

In [54]:
del gb_model, y_pred, accuracy

### 4.11 LightGBM

In [ ]:
import lightgbm as lgb
from sklearn.metrics import accuracy_score

# Membuat dataset LightGBM
train_data = lgb.Dataset(X_train, label=y_train)
test_data = lgb.Dataset(X_test, label=y_test, reference=train_data)

# Menentukan parameter LightGBM
params = {
    'objective': 'binary',  # Klasifikasi biner
    'metric': 'binary_error',  # Menggunakan binary_error sebagai metrik evaluasi
    'num_leaves': 31,  # Jumlah daun dalam setiap pohon
    'learning_rate': 0.07,  # Kecepatan pembelajaran 7
    'n_estimators': 100  # Jumlah estimators (pohon)
}

# Melatih model dengan LightGBM
gbm = lgb.train(params, train_data, valid_sets=[test_data])

# Memprediksi data uji
y_pred = gbm.predict(X_test, num_iteration=gbm.best_iteration)

# Mengubah probabilitas menjadi prediksi biner
y_pred_binary = (y_pred >= 0.5).astype(int)

# Menghitung akurasi
accuracy = accuracy_score(y_test, y_pred_binary)
print(f"Akurasi model: {accuracy * 100:.2f}%")

[LightGBM] [Info] Number of positive: 1416, number of negative: 1384
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000225 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 788
[LightGBM] [Info] Number of data points in the train set: 2800, number of used features: 13
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.505714 -> initscore=0.022858
[LightGBM] [Info] Start training from score 0.022858
Akurasi model: 91.50%


In [65]:
test_pred = gbm.predict(test_df, num_iteration=gbm.best_iteration)

test_pred_binary = (test_pred >= 0.5).astype(int)

df_test_pred = pd.DataFrame(test_pred_binary)

df_test_pred.to_csv(f"submission-{2}.csv", index=False, header=['will_buy_on_return_visit'])

In [67]:
del test_pred, test_pred_binary

In [ ]:
del train_data, test_data, params, gbm, y_pred, y_pred_binary, accuracy

## 5. Model Evaluation

| **Nama Metode**                     | **Akurasi**      | **Public Score**  |
|-------------------------------------|------------------|-------------------|
| **Model Klasik (Tradisional)**      |                  |                   |
| Logistic Regression                 | [Akurasi]        | [Public Score]    |
| K-Nearest Neighbors (KNN)           | 75.00%           | [Public Score]    |
| Naive Bayes                         |                  |                   |
| - Gaussian Naive Bayes             | 71.25%           | [Public Score]    |
| - Multinomial Naive Bayes          | 66.33%           | [Public Score]    |
| - Bernoulli Naive Bayes            | 79.25%           | [Public Score]    |
| Decision Tree                       | 86.58%           | [Public Score]    |
| Random Forest                       | 89.42%           | [Public Score]    |
| Support Vector Machine (SVM)        | [Akurasi]        | [Public Score]    |
| Linear Discriminant Analysis (LDA)  | [Akurasi]        | [Public Score]    |
| Quadratic Discriminant Analysis (QDA) | [Akurasi]       | [Public Score]    |
| **Ensemble Methods**                |                  |                   |
| Bagging Classifier                  |                  |                   |
| - Bagged Decision Trees             | [Akurasi]        | [Public Score]    |
| - Random Forest (juga termasuk ensemble) | 89.42%     | [Public Score]    |
| Boosting Algorithms                 |                  |                   |
| - AdaBoost                          | 90.67%           | [Public Score]    |
| - Gradient Boosting                 | 92.08%           | [Public Score]    |
| - XGBoost                           | 91.50%           | [Public Score]    |
| - LightGBM                          | 91.50%           | 71.71428%    |
| - CatBoost                          | 91.33%           | 68.57142%    |
| Stacking & Voting                   |                  |                   |
| - Stacking Classifier               | [Akurasi]        | [Public Score]    |
| - Voting Classifier                 | [Akurasi]        | [Public Score]    |
